dask/array/gufunc.py

```diff
@@ -0,0 +1,343 @@
+    for e in concat(ins):
+        if len(_valid_name_re.findall(e)) != 1:
+            raise ValueError('`"{0}"` is not a valid name.'.format(e))
+    ins = ",".join(("({0})".format(",".join(dims)) for dims in ins))
+
+    if outs is None:
+        outs = ""
+    elif isinstance(outs, list):
+        for e in concat(outs):
+            if len(_valid_name_re.findall(e)) != 1:
+                raise ValueError('`"{0}"` is not a valid name.'.format(e))
+        _flag = len(outs) == 1
+        outs = ",".join(("({0})".format(",".join(dims)) for dims in outs))
+        if _flag:
+            outs += ","
+    else:
+        for e in outs:
+            if len(_valid_name_re.findall(e)) != 1:
+                raise ValueError('`"{0}"` is not a valid name.'.format(e))
+        outs = "({0})".format(",".join(outs))
+    return "{0}->{1}".format(ins, outs)
+
+
+def apply_gufunc(func, signature, *args, **kwargs):
+    """
+    Apply a generalized ufunc [2]_ to arrays. The function is
+    mapped to the input arguments. The ``signature`` determines
+    if the function consumes or produces core dimensions. The
+    remaining dimensions in given input arrays (``*args``) are
+    considered loop dimensions and are required to broadcast
+    naturally against each other.
+
+    In other terms, this function is like np.vectorize, but for
+    the blocks of dask arrays. If the function itself shall also
+    be vectorized, ``vectorize=True`` can be used for convenience.
+
+    Parameters
+    ----------
+    func : callable
+        Function to call like ``func(*args, **kwargs)`` on input arrays
+        (``*args``) that returns an array or tuple of arrays. If multiple
+        arguments with non-matching dimensions are supplied, this function is
+        expected to vectorize (broadcast) over axes of positional arguments in
+        the style of NumPy universal functions [1]_ (if this is not the case,
+        set ``vectorize=True``).
+    signature: String
+        Specifies what core dimensions are consumed and produced by ``func``.
+        According to the specification of numpy.gufunc signature [2]_
+    output_dtypes : dtype or list of dtypes
+        dtype or list of output dtypes.
+    output_sizes : dict, optional
+        Optional mapping from dimension names to sizes for outputs. Only used if
+        new core dimensions (not found on inputs) appear on outputs.
+    vectorize: bool
+        If set to ``True``, ``np.vectorize`` is applied to ``func`` for
+        convenience. Defaults to ``False``.
+
+    Returns
+    -------
+    Single dask.array.Array or tuple of dask.array.Array
+
+    Examples
+    --------
+    >>> import dask.array as da
+    ... import numpy as np
+    ... def stats(x):
+    ...     return np.mean(x, axis=-1), np.std(x, axis=-1)
+    ... a = da.random.normal(size=(10,20,30), chunks=5)
+    ... mean, std = da.apply_gufunc(stats, "(i)->(),()", a, output_dtypes=2*(a.dtype,))
+    ... mean.compute()  # doctest: +SKIP
```

Comment:
Style note, I recommend using `>>>` on each individual line.

Reply:
Except the body of the stats function, of course

Reply:
Done